File: algorithms/python/austrailianvoting.py
import sys

buffer = []
for line in sys.stdin:
    line = line[:-1]
    buffer.append(line)

cases = buffer.pop(0)
current_index = 1
while current_index < len(buffer):
    candidates_count = int(buffer[current_index])
    current_index += 1
    candidates = buffer[current_index:current_index + candidates_count]
    current_index += candidates_count
    votes = []
    while buffer[current_index] != '':
        ballot = buffer[current_index].split()
        votes.append(ballot)
        current_index += 1
    tally = [0]*len(votes)
    for vote in votes:
        tally[int(vote[0])-1] += 1
    winner = ''
    excluded_candidates = []
    while winner == '':
        for index, candidate in tally:
            if candidate > len(votes/2):
                winner = candidates[index]

    print winner
    current_index += 30
print buffer
print cases

File: algorithms/javascript/austrailianvoting.js
let totalInput = ''
process.stdin.resume()
process.stdin.setEncoding('ascii')
process.stdin.on('data', (input) => {
  totalInput += input
})

process.stdin.on('end', () => {
  processData(totalInput)
})

const processData = (input) => {
  const inArr = input.split('\n')
  const cases = parseInt(inArr.shift(), 10)
  let inputIndex = 1 // skip first blank line

  for (let i = 0; i < cases; i++, inputIndex++) {
    const candidateCount = parseInt(inArr[inputIndex++], 10)
    const candidates = inArr.slice(inputIndex, inputIndex + candidateCount)
    inputIndex += candidateCount

    let activeCandidates = {}
    let votes = {}
    for (let num = 1; num <= candidates.length; num++) {
      activeCandidates[num] = true
    }

    const startOfBallots = inputIndex
    let totalBallots = 0
    let winner = ''
    while (winner === '') {
      for (let num = 1; num <= candidates.length; num++) {
        votes[num] = 0
      }
      for (let ballotCount = 0; inArr[startOfBallots + ballotCount] !== ''; ballotCount++) {
        if (ballotCount > totalBallots) totalBallots = ballotCount
        const ballot = inArr[startOfBallots + ballotCount].split(' ')
        for (let voteIndex = 0; voteIndex < ballot.length; voteIndex++) {
          const vote = parseInt(ballot[voteIndex], 10)
          if (activeCandidates[vote]) {
            votes[vote]++
            break // Successful vote
          }
        }
      }
      let worstTotal = votes['1']
      for (candidate in votes) {
        if (votes[candidate] < worstTotal) worstTotal = votes[candidate]
        if (votes[candidate] / totalBallots > 0.5) winner = candidates[candidate - 1]
      }
      for (candidate in votes) {
        if (votes[candidate] === worstTotal) activeCandidates[candidate] = false
      }
    }
    console.log(winner)
    inputIndex += totalBallots
  }
}
